hyperlink/_url.py

```diff
@@ -1464,3 +1504,329 @@ def remove(self, name):
+            path = [_encode_reserved(p) for p in path]
+        if query is not _UNSET:
+            query = [[_encode_reserved(x)
+                      if x is not None else None
+                      for x in (k, v)]
+                     for k, v in iter_pairs(query)]
+        if userinfo is not _UNSET:
+            if len(userinfo) > 2:
+                raise ValueError('userinfo expected sequence of ["user"] or'
+                                 ' ["user", "password"], got %r' % userinfo)
+            userinfo = u':'.join([_encode_reserved(p) for p in userinfo])
+        new_url = self._url.replace(scheme=scheme,
+                                    host=host,
+                                    path=path,
+                                    query=query,
+                                    fragment=fragment,
+                                    port=port,
+                                    rooted=rooted,
+                                    userinfo=userinfo,
+                                    uses_netloc=uses_netloc)
+        return type(self)(url=new_url)
+
+    def get(self, name):
+        "Get the value of all query parameters whose name matches *name*"
+        return [v for (k, v) in self.query if name == k]
+
+    def add(self, name, value=None):
+        "Return a new DecodedURL with the query parameter *name* and *value* added."
+        return self.replace(query=self.query + ((name, value),))
+
+    def set(self, name, value=None):
+        "Return a new DecodedURL with query parameter *name* set to *value*"
+        query = self.query
+        q = [(k, v) for (k, v) in query if k != name]
+        idx = next((i for (i, (k, v)) in enumerate(query) if k == name), -1)
+        q[idx:idx] = [(name, value)]
+        return self.replace(query=q)
+
+    def remove(self, name):
+        "Return a new DecodedURL with query parameter *name* removed."
+        return self.replace(query=((k, v) for (k, v) in self.query
+                                   if k != name))
+
+    def __repr__(self):
+        cn = self.__class__.__name__
+        return '%s(url=%r)' % (cn, self._url)
+
+    def __str__(self):
+        # TODO: the underlying URL's __str__ needs to change to make
+        # this work as the URL, see #55
+        return str(self._url)
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return NotImplemented
+        return self.normalize().to_uri() == other.normalize().to_uri()
+
+    def __ne__(self, other):
+        if not isinstance(other, self.__class__):
+            return NotImplemented
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.__class__, self.scheme, self.userinfo, self.host,
+                     self.path, self.query, self.fragment, self.port,
+                     self.rooted, self.uses_netloc))
+
+    # # Begin Twisted Compat Code
+    asURI = to_uri
+    asIRI = to_iri
+
+    @classmethod
```

Comment:
Why is this its own `classmethod` when the instance methods are aliased?

Reply:
Twisted-style single character argument name copied for consistency with URL.